src/lib.rs

```diff
@@ -1,55 +1,97 @@
-//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Consensus] struct
-//! requires access to an [Environment] object which black-boxes the network layer and gives
-//! appropriate access to the set of available blocks that we need to make consensus on.
+//! Implements the Aleph BFT Consensus protocol as a "finality gadget". The [Member] struct
+//! requires access to a network layer, a cryptographic primitive, and a data provided that
+//! gives appropriate access to the set of available data that we need to make consensus on.
 
 use codec::{Decode, Encode};
-use futures::{Future, Sink, Stream};
-use log::{debug, error};
+use futures::Future;
 use std::{
     fmt::{Debug, Display},
-    hash::Hash,
-};
-use tokio::{
-    sync::{mpsc, oneshot},
-    time::Duration,
+    hash::Hash as StdHash,
 };
+use tokio::sync::mpsc;
 
-use crate::{creator::Creator, extender::Extender, syncer::Syncer, terminal::Terminal};
+use crate::nodes::{NodeCount, NodeIndex, NodeMap};
 
-pub use crate::nodes::{NodeCount, NodeIndex, NodeMap};
+pub use member::{Config, Member};
 
+mod bft;
+mod consensus;
 mod creator;
 mod extender;
+mod member;
 pub mod nodes;
 mod syncer;
 mod terminal;
 mod testing;
+mod units;
+
+pub trait DataProvider<Data> {
+    type Error: Debug;
+    fn get_data(&self) -> Data;
+    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error>;
+}
+
+pub trait KeyBox<Signature>: MyIndex {
+    fn sign(&self, msg: &Vec<u8>) -> Signature;
+    fn verify(&self, msg: &Vec<u8>, sgn: &Signature) -> bool;
+}
+
+#[async_trait::async_trait]
+pub trait Network {
+    type Error: Debug;
+    fn send(&self, command: NetworkCommand) -> Result<(), Self::Error>;
```

Comment:
the `NetworkCommand` seems unnecessary here. Maybe we can replace `send` with separate methods `send_to_all`, `send_to_peer`, `send_to_rand_peer` and `broadcast_reliably`?

Reply:
The reason behind the name "NetworkCommand" is that we might want to add some other "commands" that are not necessarily "send" commands. Concretely, there might be commands to change some PeerId's reputation, and/or commands to connect/disconnect to some node. The api with multiple methods sounds good to me, we can do that, but I wouldn't push for changing it in this PR, as it's huge already.